File: Sources/windows/BitTherapyCLR/src/game/animator.cpp
#include "animator.h"

#include <chrono>
#include <vector>

Animator::Animator() : 
    initialValue(0), 
    finalValue(0), 
    currentValue(0),
    incrementPerMillisecond(0),
    animationCurve(AnimationCurve::linear),
    isActive(false) 
{}

void Animator::animate(
    double initialValue, 
    double finalValue, 
    std::chrono::milliseconds duration,
    AnimationCurve animationCurve
) {
    this->currentValue = initialValue;
    this->initialValue = initialValue;
    this->finalValue = finalValue;
    this->animationCurve = animationCurve;

    auto diff = finalValue - initialValue;
    auto durationMs = 1000 * duration.count();
    this->incrementPerMillisecond = diff / durationMs;

    this->isActive = true;
}
    
void Animator::update(std::chrono::milliseconds timeSinceLastUpdate) {
    if (!isActive) { return; }
    auto elapsedMs = 1000 * timeSinceLastUpdate.count();
    currentValue = currentValue + elapsedMs * incrementPerMillisecond;

    if (currentValue >= finalValue) {
        currentValue = finalValue;
        isActive = false;
    }
}

const double Animator::current() const {
    return currentValue;
}

File: Sources/windows/App/Sources/Game/Animator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace App
{
    public enum AnimationCurve
    {
        Linear = 0
    }

    public class Animator
    {
        private double initialValue;
        private double finalValue;
        private double currentValue;
        private double incrementPerMillisecond;
        private bool isActive;
        private AnimationCurve animationCurve;

        public Animator()
        {
            initialValue = 0;
            finalValue = 0;
            currentValue = 0;
            incrementPerMillisecond = 0;
            animationCurve = AnimationCurve.Linear;
            isActive = false;
        }

        public void Animate(double initialValue, double finalValue, TimeSpan duration, AnimationCurve animationCurve)
        {
            this.currentValue = initialValue;
            this.initialValue = initialValue;
            this.finalValue = finalValue;
            this.animationCurve = animationCurve;

            var diff = finalValue - initialValue;
            var durationMs = duration.TotalMilliseconds;
            this.incrementPerMillisecond = diff / durationMs;

            this.isActive = true;
        }

        public void Update(TimeSpan timeSinceLastUpdate)
        {
            if (!isActive) return;

            var elapsedMs = timeSinceLastUpdate.TotalMilliseconds;
            currentValue = currentValue + elapsedMs * incrementPerMillisecond;

            if (currentValue >= finalValue)
            {
                currentValue = finalValue;
                isActive = false;
            }
        }

        public double Current()
        {
            return currentValue;
        }
    }
}
